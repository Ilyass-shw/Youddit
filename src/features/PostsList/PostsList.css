.posts-container {
	display: inline-flex;
	margin-left: 4.5rem;
	background-color: rgba(24, 24, 24, 1);
	min-height: 100vh;
}
.posts-list {
	width: 100vh;
	margin-top: 1.5rem;
	flex: 1;
}
