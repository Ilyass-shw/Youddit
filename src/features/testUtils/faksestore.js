import { configureStore } from "@reduxjs/toolkit";
import { createSlice } from "@reduxjs/toolkit";

const initialState = {
	firstFilter: "relevance",
	lastFilter: "comments",
	filter: "relevance",
};
const fakeSlice = createSlice({
	name: "posts",
	initialState,
	reducers: {},
});

export default configureStore({
	reducer: {
		posts: fakeSlice.reducer,
	},
});

export const makeCustomTestStore = (name, initialState, reducers = {}) => {
	const fakeSlice = createSlice({
		name,
		initialState,
		reducers,
	});

	return configureStore({
		reducer: {
			posts: fakeSlice.reducer,
		},
	});
};
