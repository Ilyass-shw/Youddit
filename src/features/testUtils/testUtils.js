import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import store, { makeCustomTestStore } from "./faksestore";

const TestProvider = ({ store, children }) => <Provider store={store}>{children}</Provider>;

export function render(ui, { store, ...otherOpts }) {
	return rtlRender(<TestProvider store={store}>{ui}</TestProvider>, otherOpts);
}

export function makeTestStore(name, initialState, reducers) {
	if (name && initialState) {
		const Store = makeCustomTestStore(name, initialState, reducers);
		const origDispatch = Store.dispatch;
		Store.dispatch = jest.fn(origDispatch);
		return Store;
	} else {
		const origDispatch = store.dispatch;
		store.dispatch = jest.fn(origDispatch);
		return store;
	}
}

// re-export everything
export * from "@testing-library/react";
