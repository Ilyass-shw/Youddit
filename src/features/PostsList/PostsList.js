import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import Post from "../posts/Post";
import "./PostsList.css";
import { fetchPosts, selectAllPosts } from "../posts/postsSlice/postsSlice";
import PostsSkeleton from "../PostsSkeleton/PostsSkeleton";
import { nanoid } from "nanoid";

const PostsList = () => {
	const postsStatus = useSelector((state) => state.posts.postFetchingStatus);
	const postError = useSelector((state) => state.posts.error);
	const posts = useSelector(selectAllPosts);

	const dispatch = useDispatch();

	useEffect(() => {
		if (postsStatus === "idle") {
			dispatch(fetchPosts());
		}
	}, [postsStatus, dispatch]);

	let content;
	if (postsStatus === "idle" || postsStatus === "Loading") {
		content = Array(3)
			.fill()
			.map(() => {
				return <PostsSkeleton key={nanoid()} />;
			});
	} else if (postsStatus === "failed") {
		content = postError;
	}
	else {
		content = posts.map((post) => {
			return <Post key={post.id} post={post} />;
		});
	}

	return (
		<div className="posts-container">
			<div className="posts-list">{content}</div>
		</div>
	);
};

export default PostsList;
