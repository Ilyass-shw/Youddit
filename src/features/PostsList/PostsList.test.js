import React from "react";
import PostsList from "./PostsList";
import { render, makeTestStore, fireEvent, screen } from "../testUtils/testUtils";

jest.mock("../PostsSkeleton/PostsSkeleton", () => {
	return {
		__esModule: true,
		default: () => {
			return <p>PostsSkeletonMock</p>;
		},
	};
});

jest.mock("../posts/Post", () => {
	return {
		__esModule: true,
		default: ({ post }) => {
			return (
				<div rol='fake post'>
					<p>{post.title}</p>
					<p>{post.img}</p>
					<p>{post.url}</p>
					<p>{post.upvotes}</p>
					<p>{post.date}</p>
					<p>{post.author}</p>
					<p>{post.subreddit}</p>
					<p>{post.id}</p>
				</div>
			);
		},
	};
});

describe("PostsList", () => {
	it("should render first time (with posts.postFetchingStatus = idle, in store) without crashing ", () => {
		const initialState = {
			postFetchingStatus: "idle",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);

		render(<PostsList />, { store });
	});

	it("should fetch posts when rendered ", () => {
		const initialState = {
			postFetchingStatus: "idle",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);

		render(<PostsList />, { store });

		expect(store.dispatch).toHaveBeenCalledTimes(1);
		// 	// expect(fetchPosts).toHaveBeenCalledTimes(2);
		// 	// expect(store.dispatch).toHaveBeenCalledWith(fetchPosts());
	});

	it("should display skeleton and match the snapshot when while posts are not fetched yet ", () => {
		const initialState = {
			postFetchingStatus: "idle",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		const { container } = render(<PostsList />, { store });
		expect(container).toMatchSnapshot();
	});

	it("should display skeleton and match the snapshot when while posts fetching is not completed yet ", () => {
		const initialState = {
			postFetchingStatus: "Loading",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		const { container } = render(<PostsList />, { store });
		expect(container).toMatchSnapshot();
	});

	it("should display error when posts fetching fails", () => {
		const initialState = {
			postFetchingStatus: "failed",
			error: "I have failed you master!",
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		const { container } = render(<PostsList />, { store });
		expect(container).toMatchSnapshot();
	});

	it("should display posts after fetching has succeeded ", () => {
		const initialState = {
			postFetchingStatus: "succeeded",
			postList: [
				{
					title: "fake title 1",
					img: "fake img url 1",
					url: "fake link 1",
					upvotes: "10k",
					date: "1 days ago",
					author: "author 1",
					subreddit: "subreddit 1",
					id: "id 1",
				},
				{
					title: "fake title 2",
					img: "fake img url 2",
					url: "fake link 2",
					upvotes: "20k",
					date: "2 days ago",
					author: "author 2",
					subreddit: "subreddit 2",
					id: "id 2",
				},
				{
					title: "fake title 3",
					img: "fake img url 3",
					url: "fake link 3",
					upvotes: "30k",
					date: "3 days ago",
					author: "author 3",
					subreddit: "subreddit 3",
					id: "id 3",
				},
			],
		};
		const name = "posts";
		const store = makeTestStore(name, initialState);
		const { container } = render(<PostsList />, { store });
		expect(container).toMatchSnapshot();
	});
});
