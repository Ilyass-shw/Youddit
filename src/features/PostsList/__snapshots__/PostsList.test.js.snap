// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsList should display error when posts fetching fails 1`] = `
<div>
  <div
    class="posts-container"
  >
    <div
      class="posts-list"
    >
      I have failed you master!
    </div>
  </div>
</div>
`;

exports[`PostsList should display posts after fetching has succeeded  1`] = `
<div>
  <div
    class="posts-container"
  >
    <div
      class="posts-list"
    >
      <div
        rol="fake post"
      >
        <p>
          fake title 1
        </p>
        <p>
          fake img url 1
        </p>
        <p>
          fake link 1
        </p>
        <p>
          10k
        </p>
        <p>
          1 days ago
        </p>
        <p>
          author 1
        </p>
        <p>
          subreddit 1
        </p>
        <p>
          id 1
        </p>
      </div>
      <div
        rol="fake post"
      >
        <p>
          fake title 2
        </p>
        <p>
          fake img url 2
        </p>
        <p>
          fake link 2
        </p>
        <p>
          20k
        </p>
        <p>
          2 days ago
        </p>
        <p>
          author 2
        </p>
        <p>
          subreddit 2
        </p>
        <p>
          id 2
        </p>
      </div>
      <div
        rol="fake post"
      >
        <p>
          fake title 3
        </p>
        <p>
          fake img url 3
        </p>
        <p>
          fake link 3
        </p>
        <p>
          30k
        </p>
        <p>
          3 days ago
        </p>
        <p>
          author 3
        </p>
        <p>
          subreddit 3
        </p>
        <p>
          id 3
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`PostsList should display skeleton and match the snapshot when while posts are not fetched yet  1`] = `
<div>
  <div
    class="posts-container"
  >
    <div
      class="posts-list"
    >
      <p>
        PostsSkeletonMock
      </p>
      <p>
        PostsSkeletonMock
      </p>
      <p>
        PostsSkeletonMock
      </p>
    </div>
  </div>
</div>
`;

exports[`PostsList should display skeleton and match the snapshot when while posts fetching is not completed yet  1`] = `
<div>
  <div
    class="posts-container"
  >
    <div
      class="posts-list"
    >
      <p>
        PostsSkeletonMock
      </p>
      <p>
        PostsSkeletonMock
      </p>
      <p>
        PostsSkeletonMock
      </p>
    </div>
  </div>
</div>
`;
